feat(mobile): use efficient sync

fyfrey opened the pull request:
follow-up for #8755

enable the use of the more efficient endpoints to perform sync operations.

only merge after #9189 has been in a release for 1-2 weeks.

fyfrey committed:
feat(mobile): use efficient sync

review feedback

alextran1502 commented on mobile/lib/shared/services/asset.service.dart:
```diff
@@ -83,38 +91,30 @@ class AssetService {
-      for (int i = 0;; i += chunkSize) {
+      DateTime? lastCreationDate;
+      String? lastId;
+      for (;;) {
```
`while` loop?

alextran1502 replied:
I see this a lot in embedded system code, not used to seeing it in higher level language 😝

fyfrey replied:
I personally use this for loops when I break out using other means.

While(true).. I use for endless tasks. Like a server listening indefinitely until it's killed etc.

martyfuhry replied:
Can this be rewritten to a `do while` loop on the condition `assets.length < chunkSize`? I think that's the only break condition I see here. I concur that `while (true)` loops are odd and potentially dangerous. I also think `for (;;)` is less readable. I did not know what it meant until I read some of the other comments.

fyfrey replied:
sadly, I can't put the break condition in the loop condition check: it would not even run the first pass of the loop - i can only to while(true) instead of for(;;). The loop has two break conditions that cannot be handled in the loop header

martyfuhry reviewed (commented):
I wish I could be more help, as I'm not tremendously familiar with the backup sync. But I'm happy to do any testing if you'd like me to, as well!

alextran1502 commented on mobile/test/modules/shared/sync_service_test.dart:
```diff
@@ -75,8 +75,12 @@ void main() {
       expect(db.assets.countSync(), 5);
-      final bool c1 =
-          await s.syncRemoteAssetsToDb(owner, _failDiff, (u) => remoteAssets);
+      final bool c1 = await s.syncRemoteAssetsToDb(
```
I agree with Marty here, I found a lot of these method with passed in function rather hard to read and follow

alextran1502 commented:
I ran into this error on the server when running the PR

![image](https://github.com/immich-app/immich/assets/27055614/02383532-6ccc-4f81-90af-b24b548a89e1)

fyfrey commented:
> I ran into this error on the server when running the PR

It seems some more server changes are required for stacked assets (my test instance probably had no stacked assets at the time...). I converted this PR back to a draft and will work on a new PR for the required server changes.

fyfrey committed:
adapt to changed  server endpoints

fyfrey committed:
formatting

jrasm91 reviewed (approved):
LGTM

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat/app-efficient-sync

alextran1502 commented:
Hi @fyfrey I tested this again and found a problem with rendering memory lane. Here is the error message

![image](https://github.com/immich-app/immich/assets/27055614/43359110-043c-4b85-a1ec-8babee9c1b3d)

From my debugging, the API returned the correct data. Still, the call to get the asset from the database when building the model probably happened before the synchronization process finished, so it could not parse all data correctly, leading to the error.

![image](https://github.com/immich-app/immich/assets/27055614/25d68fb6-fc8f-4356-b7a2-430072fa8842)

fyfrey committed:
fix memory lane bug

fyfrey commented:
> Hi @fyfrey I tested this again and found a problem with rendering memory lane. Here is the error message

Thanks @alextran1502 ! I've pushed a new commit to fix the bug in the memory lane. We were just lucky it did not occur previously..
Solution: Do not add a memory lane if assets are not yet in DB. Refresh memory lane provider after remote asset sync.

alextran1502 commented:
Hi @fyfrey, the change doesn't entirely fix the issue. I have multiple memories on a day but only the first memory gets rendered, so I checked the database. On `main`, I see 69000 assets in Isar database, while on this branch, I only see 64000 assets, so the fetching mechanism somehow doesn't fetch all assets

fyfrey commented:
> Hi @fyfrey, the change doesn't entirely fix the issue. I have multiple memories on a day but only the first memory gets rendered, so I checked the database. On `main`, I see 69000 assets in Isar database, while on this branch, I only see 64000 assets, so the fetching mechanism somehow doesn't fetch all assets

Thank you for testing again @alextran1502 . Very interesting. I did not notice any sync discrepancy (asset count) with my initial server changes. In the newer fix PR (#9100) we changed it to use the common asset query builder. It turns out it is incompatible with the app with stacked assets as the app expects all assets individually while the server returns them in an array with of the  parent asset. This is probably the same issue you encountered a while ago and reverted server refactoring changes in #7752.

made another server PR to fix it: #9189

alextran1502 commented:
I tried again with #9189 on my prod (FOR SCIENCE!!) and I still see missing memories. Checking the database again, now I see 68000 assets are fetched on #9189 while still 69000 assets on `main`

fyfrey commented:
Something must still be different between the old and the new server query.
We need some internal test instance that has all the edge cases like your prod instance does :-P

The old method also fetched archived assets for partners, maybe this accounts for the difference. But memories should not show partner assets.... I'll take another look at the code, or better the generated SQL

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat/app-efficient-sync

alextran1502 committed:
fix: bad merge

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat/app-efficient-sync

alextran1502 committed:
fix call not returning correct number of asset
